### PR TITLE
fix(gpu): general fixes and improvements to PBS

### PR DESCRIPTION
**Pull request by pdroalves**
closes: https://github.com/zama-ai/tfhe-rs-internal/issues/743, https://github.com/zama-ai/tfhe-rs-internal/issues/811

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Review by agnesLeroy — COMMENTED** [review]
Thanks a lot for this @pdroalves! It would be good to make the same changes in this PR for the tbc variant as well I think, to keep things consistent between cg & tbc.

File: backends/tfhe-cuda-backend/cuda/include/pbs/pbs_multibit_utilities.h [hunk 1]
```diff
@@ -225,10 +225,12 @@ template <typename Torus> struct pbs_buffer<Torus, PBS_TYPE::MULTI_BIT> {
           num_blocks_keybundle * (polynomial_size / 2) * sizeof(double2),
           stream, gpu_index);
       global_accumulator = (Torus *)cuda_malloc_async(
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/include/pbs/pbs_multibit_utilities.h:227]: I think this global_accumulator needs to be kept as is for the "default" multibit PBS variant. Only for TBC and CG it should not be sized with levels, am I wrong?

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap.cuh [hunk 2]
```diff
@@ -113,7 +113,7 @@ mul_ggsw_glwe(Torus *accumulator, double2 *fft, double2 *join_buffer,
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap.cuh:113]: In this file, this section:
```
  // accumulate rest of the products into fft buffer
  for (int l = 1; l < gridDim.x; l++) {
    auto cur_src_acc = get_join_buffer_element<G>(l, blockIdx.y, group,
                                                  join_buffer, polynomial_size,
                                                  glwe_dimension, support_dsm);
    tid = threadIdx.x;
    for (int i = 0; i < params::opt / 2; i++) {
      fft[tid] += cur_src_acc[tid];
      tid += params::degree / params::opt;
```
could be done on blocks whose blockIDx.x = 0 only as well no?

File: backends/tfhe-cuda-backend/cuda/include/pbs/pbs_multibit_utilities.h (hunk 1 above)
Comment (pdroalves) [reply backends/tfhe-cuda-backend/cuda/include/pbs/pbs_multibit_utilities.h:227]: Ah yes, I just renamed it. It is used as the "join buffer" in `mul_ggsw_glwe`, so I'm calling it by the right name.

**Review by agnesLeroy — APPROVED** [review]
Thanks a lot for this @pdroalves! I believe we still have an issue with the cmux test though no?

**Comment by pdroalves** [comment]
Performance holds for this branch.

**Pull request merged**